0012-five-bells-ledger-api/0012-five-bells-ledger-api.md

```diff
@@ -0,0 +1,1290 @@
+
+{
+  "from": "https://blue.ilpdemo.ripple.com/ledger/accounts/bob",
+  "to": "https://blue.ilpdemo.ripple.com/ledger/accounts/alice",
+  "data": {
+    "method": "quote_request",
+    "id": "721e4126-98a1-4974-b35a-8a8f4655f934",
+    "data": {
+      "source_amount": "100.25",
+      "source_address": "example.eur-ledger.alice",
+      "destination_address": "example.usd-ledger.bob",
+      "source_expiry_duration": "6000",
+      "destination_expiry_duration": "5"
+    }
+  }
+}
+```
+
+Response:
+
+```
+HTTP/1.1 204 No Content
+```
+
+#### Errors
+
+- [InvalidBodyError][] - The request was not valid JSON, or did not match the schema for the request.
+- [InvalidUriParameterError][] - One of the URLs, URIs, or UUIDs in the body is not validly formatted.
+- [UnprocessableEntityError][] - The object was syntactically valid but had a semantic issue.
+
+
+### Get Auth Token
+[Get Auth Token]: #get-auth-token
+
+Get a token that can be used to authenticate future requests.
+
+**Note:** This method is REQUIRED for ledgers to authenticate [WebSocket][] connections. If the ledger does not authenticate WebSocket connections, this method is OPTIONAL. The ledger MAY allow clients to authenticate for any other methods of the API using the tokens returned by this method.
+
+#### Request Format
+
+```
+GET /auth_token
+```
+
+Depending on the [authentication mechanism](#authorization-and-authentication) used by this ledger, the ledger MAY require the HTTP `Auth` header with a valid username and password, or the ledger may require a different method of authentication.
+
+**[Metadata URL Name][]:** `auth_token`.
+
+#### Response Format
+
+A successful response uses the HTTP response code **200 OK** and contains a JSON object with a token that can be used for subsequent requests. The ledger can generate the token with [JWT](https://jwt.io/) or any similar system.
+
+#### Example
+
+Request:
+
+```
+GET /auth_token
+Authorization: Basic myUsername:securePassphrase
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "token": "9AtVZPN3t49Kx07stO813UHXv6pcES"
+}
+```
+
+#### Errors
+
+- [Unauthorized][]
+
+
+
+
+## WebSocket
+[WebSocket]: #websocket
+
+**[Metadata URL Name][]:** `websocket`. Suggested generic path: `/websocket` (with `wss://` protocol)
+
+Clients can subscribe to live, read-only notifications of ledger activity by opening a [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) connection to this path and sending a subscription request.
+
+### WebSocket Authentication
+
+Clients MUST authenticate themselves with a token from the [Get Auth Token][] method. This token can be supplied in one of two ways:
+
+- In the `token` query parameter. For example:
+
+        wss://red.ilpdemo.org/ledger/websocket/?token=9AtVZPN3t49Kx07stO813UHXv6pcES
+
+- As a "Bearer" token in the `Authorization` HTTP header of the request (before the protocol switch to WebSocket). For example:
```

Comment:
WebSocket Authentication is a bit tricky. This is what the RFC says:
> This protocol doesn't prescribe any particular way that servers can authenticate clients during the WebSocket handshake. The WebSocket server can use any client authentication mechanism available to a generic HTTP server, such as cookies, HTTP authentication, or TLS authentication. [RFC6455](https://tools.ietf.org/html/rfc6455)

There are a couple of options to authenticate clients. Which one to choose depends on the browser applications we want to support, so this should be an open discussion. Some options are:

1. **Cookie-based**: When the browser calls the endpoint `/auth_token` on the ledger, the ledger sets a cookie. The ledger does this by adding an HTTP Header to the response, e.g.: `Set-Cookie: token=some-json-web-token; HttpOnly; Path=/websocket`. When the client-side javascript opens the websocket (e.g. `new WebSocket('ws://ledger.example:3000/websocket')`), the browser automatically appends the cookie to the HTTP upgrade request, e.g.:
```
GET /websocket HTTP/1.1
User-Agent: Mozilla/5.0 ...
...
Cookie: token=some-json-web-token // auth details go here
...
Upgrade: websocket
```

From a security standpoint this option has the advantage that the token is not exposed to client-side javascript and, hence, xss vulnerabilities can't be used to steal your auth token.

2. **Websocket Subprotocol**: When opening a WebSocket with javascript, it is possible to indicate a websocket subprotocol. We could use this subprotocol to communicate auth details. For example, when client-side js opens a WebSocket like this `new WebSocket('ws://ledger.example:3000/websocket', 'token=somejsonwebtoken')`, the resulting HTTP request contains this HTTP Header: `Sec-WebSocket-Protocol: token=somejsonwebtoken`.

This option feels a bit like a hack, since the subprotocol header is not dedicated to transport auth details.

3. **Once the websocket is open**: Once the websocket connection is established, the client sends first a message with his credentials.

4. **Query String**: We could send the auth token in the query string, e.g. `new WebSocket('ws://ledger.example:3000/websocket?token=somejsonwebtoken')`. However, I think this is the least secure option. Refer to the [OAuth RFC](https://tools.ietf.org/html/rfc6750#section-5) for some more insights, quote:

> Because of the security weaknesses associated with the URI method   (see Section 5), including the high likelihood that the URL containing the access token will be logged, it SHOULD NOT be used unless it is impossible to transport the access token in the   “Authorization” request header field or the HTTP request entity-body.

Slack uses a creative approach to mitigate this problem. If a client wants to connect, they give out a "single-use" WebSocket URL that is only valid for 30 seconds. We could  easily copy this approach by expiring the auth tokens that the ledger gives out after 30 seconds. So if the token ends up being logged by a proxy, the token is useless after a short time period.